libgnucash/engine/Account.cpp

```diff
@@ -5417,120 +5419,155 @@ gnc_account_imap_delete_account (GncImportMatchMap *imap,
-         * NOTE: so we only keep track of a running product(A*B*C...)
-         * and product difference ((1-A)(1-B)...)
-         */
-        int32_t probability = (account_probability.product /
-                (account_probability.product + account_probability.product_difference)) * probability_factor;
-        ret.push_back({first_pass_prob.first, probability});
+        /* query token frequencies from import match map */
+        TokenAccountsInfo token_info{};
+        std::string token_name{static_cast <char const*> (current_token->data)};
+        auto path = std::string(IMAP_FRAME_BAYES "/") + token_name + "/";
+        qof_instance_foreach_slot_prefix (QOF_INSTANCE (imap->acc), path, &build_token_info, token_info);
+
+        /* calculate conditional account probabilities P(A|T) of account A given token T */
+        if (token_info.total_count > 0)
+        {
+            for (auto & account : token_info.accounts)
+                account.probability = (double)account.token_count / (double)token_info.total_count;
+            ret.emplace_back(token_name, token_info);
+        }
     }
     return ret;
 }
 
-static AccountInfo
-highest_probability(FinalProbabilityVec const & probabilities)
+/** Threshold for conditional account probabilities given individual tokens */
+static constexpr double token_account_probability_threshold = .90; /* 90% */
+
+static TokenSelection
+select_tokens(TokenInfoVec const & token_info)
 {
-    AccountInfo ret {"", std::numeric_limits<int32_t>::min()};
-    for (auto const & prob : probabilities)
-        if (prob.second > ret.probability)
-            ret = AccountInfo {prob.first, prob.second};
+    TokenSelection ret;
+    for (auto const & token : token_info)
+    {
+        /* select tokens with at least one conditional probability P(A|T)
+         * larger than token_account_probability_threshold */
+        auto const & accounts = token.second.accounts;
+        auto use_token = std::accumulate(accounts.begin(), accounts.end(), false,
+            [](bool const& a, AccountTokenProbability const& v){
+                return a | (v.probability > token_account_probability_threshold);
+            });
+        if (use_token)
+            ret.emplace(token.first);
+    }
     return ret;
 }
 
-static ProbabilityVec
-get_first_pass_probabilities(GncImportMatchMap * imap, GList * tokens)
+/** We scale the probability values by probability_factor.
+  ie. with probability_factor of 100000, 10% would be
+  0.10 * 100000 = 10000 */
+static constexpr int probability_factor = 100000;
+
+static FinalProbabilityVec
+build_probabilities(TokenInfoVec const & token_info, TokenSelection token_selection)
 {
-    ProbabilityVec ret;
-    /* find the probability for each account that contains any of the tokens
-     * in the input tokens list. */
-    for (auto current_token = tokens; current_token; current_token = current_token->next)
+    std::map<std::string, double> sum_of_probabilities;
```

Comment:
Can this be `std::unordered_map`?

Reply:
Unless there's a really good reason not to, it should be `std::vector`.

Reply:
The reason for using a map here is, that this avoids an explicit search for existing entries. This is implicitly done by function `std::map::emplace()` (see line 5545). That's a good reason for me to prefer a map to `std::vector`.

@bhardwajs What improvement do you have in mind, when suggesting `std::unordered_map`?

Reply:
Complexity of `unordered_map::emplace` is amortized constant on average while that of `map::emplace` is logarithmic in the size of the container. If the container needs to have unique collection of key-value pairs, my preference is to use `unordered_map` unless the container needs to be sorted
https://en.cppreference.com/w/cpp/container/unordered_map/emplace
https://en.cppreference.com/w/cpp/container/map/emplace

Reply:
More usefully, `std::map` builds a b-tree of nodes (hence the log<sub>2</sub> complexity) while `std::unordered_map` is more like `GHashTable`: It hashes the keys and makes a linked list of entries on each hash value. It has been repeatedly demonstrated that linear search on a std::vector is several times faster than indexed lookup on any other container, [an example](https://blog.quasardb.net/using-c-containers-efficiently) and several orders of magnitude faster than std::map and std::unordered_map. I've heard claims that [Abseil's `flat_hash_map`](https://abseil.io/docs/cpp/guides/container) is nearly as fast as std::vector but I've not seen any benchmarks and I'd rather not add the dependency.